analyses/cnv-chrom-plot/util/bin-coverage.R

```diff
@@ -0,0 +1,142 @@
+  #               The `biospecimen slot will be used to split out the `sample_id`'s corresponding ranges.   
+  #               The `status` slot should have gain/loss/neutral. 
+  #   bin_ranges: A binned GenomicRanges made from tileGenome that has been uncompressed with `unlist`. 
+  #   perc_delta_threshold: What covereage percent difference do you need to make the call?
+  #
+  # Returns:
+  #  a small data.frame that contains the status call of the sample for each bin. 
+  #
+  # Extract the ranges for this sample
+  sample_seg_ranges <- seg_ranges[which(seg_ranges$biospecimen == sample_id)]
+  
+  # Split ranges into their respective statuses
+  gain_ranges <- sample_seg_ranges[sample_seg_ranges$status == "gain"]
+  loss_ranges <- sample_seg_ranges[sample_seg_ranges$status == "loss"]
+  neutral_ranges <- sample_seg_ranges[sample_seg_ranges$status == "neutral"]
+  
+  # Calculate length of each type of status per bin
+  gain_per_bin <- bp_per_bin(bin_ranges, gain_ranges)
+  loss_per_bin <- bp_per_bin(bin_ranges, loss_ranges)
+  neutral_per_bin <- bp_per_bin(bin_ranges, neutral_ranges)
+  
+  # Format this data into one data.frame where each row is a bin
+  bin_bp_status <- data.frame(
+    bin = as.numeric(1:length(bin_ranges)),
+    # Keep bin width
+    bin_width = bin_ranges@ranges@width
+  ) %>%
+    # Join gains coverage data
+    dplyr::left_join(gain_per_bin,
+                     by = "bin"
+    ) %>%
+    # Join loss coverage data
+    dplyr::left_join(loss_per_bin,
+                     by = "bin",
+                     suffix = c(".gain", ".loss")
+    ) %>%
+    # Join neutral coverage data
+    dplyr::left_join(neutral_per_bin,
+                     by = "bin"
+    ) %>%
+    # If there is an NA, at this point we can assume it means 0
+    dplyr::mutate_at(
+      dplyr::vars(
+        dplyr::starts_with("bp_per_bin")
+      ),
+      ~ tidyr::replace_na(., 0)
+    ) %>%
+    # Reformat neutral so it is like the others
+    dplyr::rename(bp_per_bin.neutral = bp_per_bin) %>%
+    # Calculate the bins percentage of each status
+    dplyr::mutate(
+      perc_gain = bp_per_bin.gain / bin_width,
+      perc_loss = bp_per_bin.loss / bin_width,
+      perc_neutral = bp_per_bin.neutral / bin_width
+    ) %>%
+    # Use these percentages for declaring final call per bin based on
+    # the perc_delta_threshold
+    dplyr::mutate(
+      status = dplyr::case_when(
+        (perc_gain - perc_loss) > perc_delta_threshold ~ "gain",
+        (perc_loss - perc_gain) > perc_delta_threshold ~ "loss",
+        TRUE ~ "neutral"
+      )
+    )
+  
+  # Format this data as a status
+  status_df <- bin_bp_status %>%
+    # Only keep the bin and status columns
+    dplyr::select(bin, status) %>%
+    # Spread this data so we can make it a sample x bin matrix later
+    tidyr::spread(bin, status) %>%
+    # Sort the bins data/columns to be in numeric order
+    dplyr::select(order(as.numeric(colnames(.))))
```

Comment:
I *think* if you do use `arrange()` before the `spread()` you can do this a bit more cleanly.